src/app/source-file-tabs/source-file-tabs.service.ts

```diff
@@ -24,6 +32,41 @@ export class SourceFileTabsService {
     return this.workflowsService.getWorkflowVersionsSourcefiles(workflowId, versionId);
   }
 
+  convertSourceFilesToFileTabs(sourcefiles: SourceFile[]):  Map<string, SourceFile[]> {
+    let fileTabs = new Map<string, SourceFile[]>();
+    if (!sourcefiles || sourcefiles.length === 0) {
+      return fileTabs;
+    }
+    this.fileTabsSchematic.forEach(fileTab => {
+      fileTab.fileTypes.forEach(fileType => {
+        const sourceFilesMatchingType = sourcefiles.filter(sourcefile => sourcefile.type === fileType);
+        if (sourceFilesMatchingType.length > 0) {
+          if (fileTabs.has(fileTab.tabName)) {
+            fileTabs.set(fileTab.tabName, fileTabs.get(fileTab.tabName).concat(sourceFilesMatchingType));
+          } else {
+            fileTabs.set(fileTab.tabName, sourceFilesMatchingType);
+          }
+        }
+      });
+    });
+    return fileTabs;
+  }
+
+  /**
+   * Given the sourcefiles in the current tab, return the validation message object that should be shown
+   */
+  getValidationMessage(sourcefiles: SourceFile[], version: WorkflowVersion): Map<string, string> {
+    let validationMessage = null;
+    version.validations.forEach((validation: Validation) => {
+      sourcefiles.forEach(file => {
```

Comment:
Bike-shedding, especially given the typical small number of files in play, but you could call `find()` and avoid unnecessary loops. Unless you're looking to grab the last validation message instead of the first one.

Reply:
No idea, copied from previous